File: electron/job/updateAppRunningStatusJob.ts
/**
 * 定时更新应用运行状态任务
 */
import * as globalStatusService from '../service/globalStatusService'
import * as localRegService from '../service/localRegService'
import * as steamAppService from '../service/steamAppService'
import * as useAppRecordService from '../service/useAppRecordService'

let intervalId: NodeJS.Timeout | null = null
let intervalTime = 5000 // 默认5秒
let isRunning = false
let lastUpdateTime = 0
let lastRunningApps: number[] = [] // 上一次运行的应用列表

/**
 * 启动定时更新任务
 */
export function startUpdateAppRunningStatusJob() {
  if (isRunning) {
    console.warn('[Job] 应用运行状态更新任务已在运行')
    return
  }

  console.warn(`[Job] 启动应用运行状态更新任务，更新间隔: ${intervalTime}ms`)
  isRunning = true

  // 立即执行一次
  updateAppRunningStatus()

  // 启动定时任务
  intervalId = setInterval(() => {
    updateAppRunningStatus()
  }, intervalTime)
}

/**
 * 停止定时更新任务
 */
export function stopUpdateAppRunningStatusJob() {
  if (!isRunning) {
    console.warn('[Job] 应用运行状态更新任务未在运行')
    return
  }

  console.warn('[Job] 停止应用运行状态更新任务')
  if (intervalId) {
    clearInterval(intervalId)
    intervalId = null
  }
  isRunning = false
}

/**
 * 设置更新间隔时间（毫秒）
 */
export function setUpdateInterval(interval: number) {
  if (interval < 1000) {
    console.warn('[Job] 更新间隔时间不能小于1000ms，已自动设置为1000ms')
    interval = 1000
  }

  intervalTime = interval
  console.warn(`[Job] 应用运行状态更新间隔已设置为: ${intervalTime}ms`)

  // 如果任务正在运行，重启以应用新的间隔时间
  if (isRunning) {
    stopUpdateAppRunningStatusJob()
    startUpdateAppRunningStatusJob()
  }
}

/**
 * 获取当前任务状态
 */
export function getJobStatus() {
  return {
    isRunning,
    intervalTime,
    lastUpdateTime,
  }
}

/**
 * 执行更新操作
 */
async function updateAppRunningStatus() {
  try {
    // 读取当前运行的应用列表
    const currentRunningApps = await localRegService.readRunningAppsReg()

    // 比较新旧值，检测变化
    const added = currentRunningApps.filter(appId => !lastRunningApps.includes(appId))
    const removed = lastRunningApps.filter(appId => !currentRunningApps.includes(appId))

    // 只在有变化时才更新数据库
    if (added.length > 0 || removed.length > 0) {
      console.warn(`[Job] 检测到运行应用变化: 新增 ${added.length} 个, 移除 ${removed.length} 个`)

      // 获取当前活跃用户的 SteamID
      const globalStatus = await globalStatusService.refreshGlobalStatus()
      const activeSteamId = globalStatus?.activeUserSteamId

      await steamAppService.refreshSteamAppInfo(globalStatus.steamPath!)

      // 更新应用运行状态
      await steamAppService.updateAppRunningStatus(currentRunningApps)

      if (activeSteamId) {
        // 记录新增的应用
        for (const appId of added) {
          await useAppRecordService.startRecord(activeSteamId, appId)
        }

        // 结束移除的应用
        for (const appId of removed) {
          await useAppRecordService.stopRecord(activeSteamId, appId)
        }
      }
      else {
        console.warn('[Job] 未找到活跃用户 SteamID，跳过记录应用使用')
      }

      // 更新上一次的运行应用列表
      lastRunningApps = currentRunningApps
    }

    // 更新上一次的检测更新时间
    lastUpdateTime = Date.now()
  }
  catch (error) {
    console.error('[Job] 应用运行状态更新失败:', error)
  }
}

File: ElectronNet/ElectronNet/Jobs/UpdateAppRunningStatusJob.cs
using ElectronNet.Services;
using Microsoft.Extensions.Hosting;
using Microsoft.Extensions.Logging;
using SteamStat.Contracts.Ipc;
using SteamStat.Core.Platform;
using SteamStat.Core.Settings;

namespace ElectronNet.Jobs;

/// <summary>
/// 定时更新应用运行状态任务
/// </summary>
public sealed class UpdateAppRunningStatusJob(
    ISteamInstallLocator installLocator,
    GlobalStatusService globalStatusService,
    SteamAppService steamAppService,
    UseAppRecordService useAppRecordService,
    TimeProvider timeProvider,
    ILogger<UpdateAppRunningStatusJob> logger) : BackgroundService, IAppRunningStatusJobController
{
    private readonly object _sync = new();
    private CancellationTokenSource _scheduleChanged = new();
    private HashSet<int> _lastRunningApps = [];
    private TimeSpan _intervalTime = TimeSpan.FromSeconds(5);
    private bool _isRunning;
    private long _lastUpdateTime;
    private int _disposed;

    public long LastUpdateTime => Interlocked.Read(ref _lastUpdateTime);

    /// <summary>
    /// 获取该定时任务相关状态
    /// </summary>
    public UpdateAppRunningStatusJobStatusDto GetStatus()
    {
        lock (_sync)
        {
            return new UpdateAppRunningStatusJobStatusDto(
                _isRunning,
                LastUpdateTime,
                _intervalTime.TotalSeconds);
        }
    }

    public void SetEnabled(bool enabled)
    {
        lock (_sync)
        {
            if (_isRunning == enabled) return;
            _isRunning = enabled;
        }
        SignalScheduleChanged();
        logger.LogInformation("Application running-status monitoring is {State}", enabled ? "enabled" : "disabled");
    }

    /// <summary>
    /// 设置更新间隔时间
    /// </summary>
    public void SetInterval(TimeSpan interval)
    {
        if (interval.TotalMilliseconds < 1000)
        {
            logger.LogWarning("应用运行状态更新间隔不能小于 1000ms，已自动设置为 1000ms");
            interval = TimeSpan.FromMilliseconds(1000);
        }

        lock (_sync)
        {
            if (_intervalTime == interval) return;
            _intervalTime = interval;
        }
        SignalScheduleChanged();
        logger.LogInformation("应用运行状态更新间隔已设置为 {IntervalMilliseconds}ms", interval.TotalMilliseconds);
    }

    public override async Task StopAsync(CancellationToken cancellationToken)
    {
        SignalScheduleChanged();
        await base.StopAsync(cancellationToken).ConfigureAwait(false);
    }

    protected override async Task ExecuteAsync(CancellationToken stoppingToken)
    {
        while (!stoppingToken.IsCancellationRequested)
        {
            bool enabled;
            TimeSpan interval;
            CancellationTokenSource iteration;
            lock (_sync)
            {
                enabled = _isRunning;
                interval = _intervalTime;
                iteration = CancellationTokenSource.CreateLinkedTokenSource(stoppingToken, _scheduleChanged.Token);
            }

            using (iteration)
            try
            {
                if (!enabled)
                {
                    await Task.Delay(Timeout.InfiniteTimeSpan, timeProvider, iteration.Token).ConfigureAwait(false);
                    continue;
                }

                await UpdateAsync(iteration.Token).ConfigureAwait(false);
                using var timer = new PeriodicTimer(interval, timeProvider);
                while (await timer.WaitForNextTickAsync(iteration.Token).ConfigureAwait(false))
                    await UpdateAsync(iteration.Token).ConfigureAwait(false);
            }
            catch (OperationCanceledException) when (iteration.IsCancellationRequested)
            {
            }
        }
    }

    /// <summary>
    /// 执行更新操作
    /// </summary>
    private async Task UpdateAsync(CancellationToken cancellationToken)
    {
        try
        {
            // 读取当前运行的应用列表
            var currentRunningApps = installLocator.ReadAppRegistry()
                .Where(a => a.Value.Running is 1)
                .Select(a => a.Key)
                .ToHashSet();

            HashSet<int> previous;
            lock (_sync) previous = [.. _lastRunningApps];
            var added = currentRunningApps.Except(previous).ToList();
            var removed = previous.Except(currentRunningApps).ToList();

            // 只在有变化时才更新数据库
            if (added.Count > 0 || removed.Count > 0)
            {
                logger.LogInformation("检测到运行应用变化: 新增 {AddedCount} 个, 移除 {RemovedCount} 个", added.Count, removed.Count);
                var globalStatus = await globalStatusService.SyncAndGetOne(log: false, cancellationToken);
                var activeSteamId = globalStatus?.ActiveUserSteamId;

                if (activeSteamId != null)
                {
                    await steamAppService.SyncDb(log: false, cancellationToken);
                    await steamAppService.UpdateAppRunningStatus(added, isRunning: true, cancellationToken);
                    await steamAppService.UpdateAppRunningStatus(removed, isRunning: false, cancellationToken);

                    foreach (var appId in added)
                        await useAppRecordService.StartRecord(activeSteamId, appId, cancellationToken);
                    foreach (var appId in removed)
                        await useAppRecordService.StopRecord(activeSteamId, appId, cancellationToken);
                }
                else
                {
                    logger.LogWarning("未找到活跃用户 SteamID，跳过记录应用使用");
                }

                lock (_sync) _lastRunningApps = currentRunningApps;
            }

            Interlocked.Exchange(ref _lastUpdateTime, timeProvider.GetUtcNow().ToUnixTimeSeconds());
        }
        catch (OperationCanceledException) when (cancellationToken.IsCancellationRequested)
        {
        }
        catch (Exception exception)
        {
            logger.LogError(exception, "应用运行状态更新失败");
        }
    }

    private void SignalScheduleChanged()
    {
        CancellationTokenSource previous;
        lock (_sync)
        {
            previous = _scheduleChanged;
            _scheduleChanged = new CancellationTokenSource();
        }
        previous.Cancel();
        previous.Dispose();
    }

    public override void Dispose()
    {
        if (Interlocked.Exchange(ref _disposed, 1) != 0) return;
        CancellationTokenSource schedule;
        lock (_sync) schedule = _scheduleChanged;
        schedule.Cancel();
        schedule.Dispose();
        base.Dispose();
    }
}
